Dockerfile

```diff
@@ -52,41 +54,84 @@ RUN ln -s /usr/bin/python3 /usr/bin/python
 
 WORKDIR /edx/app/edx-platform/edx-platform
 
-COPY . /edx/app/edx-platform/edx-platform
-
 ENV PATH /edx/app/edx-platform/nodeenv/bin:${PATH}
 ENV PATH ./node_modules/.bin:${PATH}
 ENV CONFIG_ROOT /edx/etc/
 ENV PATH /edx/app/edx-platform/edx-platform/bin:${PATH}
 ENV SETTINGS production
+RUN mkdir -p /edx/etc/
 
-# TODO: Install requirements before copying in code.
+# Distinguish requirements that come from code within edx-platform
+# from external requirements (specifically, those from pip or GitHub).
+# Then, install just external requirements.
+# This way, the cached installations of external requirements are not
+# busted whenever edx-platform code is changed.
+# TODO this is quite a hack and should be done in `make upgrade` probably.
 RUN pip install setuptools==39.0.1 pip==9.0.3
-RUN pip install -r requirements/edx/base.txt
+COPY requirements/ requirements/
+RUN grep -E    "^-e (common|openedx|lms|cms|\.)(/| )" requirements/edx/base.txt > requirements/edx/base_in_tree.txt
+RUN grep -E -v "^-e (common|openedx|lms|cms|\.)(/| )" requirements/edx/base.txt > requirements/edx/base_not_in_tree.txt
+RUN pip install -r requirements/edx/base_not_in_tree.txt
 
+# Copy just JS requirements and install them.
+COPY package.json package.json
+COPY package-lock.json package-lock.json
+# TODO: Shouldn't we use node==12.11.1?
 RUN nodeenv /edx/app/edx-platform/nodeenv --node=8.9.3 --prebuilt
```

Comment:
@dianakhuang pointed out that we're on a higher Node version now. The `--node=` should be updated to something in `12.x.x`.